pennylane/spin/lattice.py

```diff
@@ -0,0 +1,196 @@
+        neighbour_order=1,
+        distance_tol=1e-5,
+    ):
+
+        self.L = np.asarray(L)
+        self.n_dim = len(L)
+        self.unit_cell = np.asarray(unit_cell)
+        if basis is None:
+            basis = np.zeros(self.unit_cell.shape[0])[None, :]
+
+        self.basis = np.asarray(basis)
+        self.n_sl = len(self.basis)
+        self.n_sites = np.prod(L) * self.n_sl
+
+        if isinstance(boundary_condition, bool):
+            boundary_condition = [boundary_condition for _ in range(self.n_dim)]
+
+        self.boundary_condition = boundary_condition
+        self._test_input_accuracy()
+
+        self.lattice_points, lattice_map = self._generate_grid(neighbour_order)
+
+        cutoff = neighbour_order * np.linalg.norm(self.unit_cell, axis=1).max() + distance_tol
+        edges = self._identify_neighbours(cutoff)
+        self._generate_true_edges(edges, lattice_map, neighbour_order)
+
+    def _test_input_accuracy(self):
+        r"""Tests the accuracy of the input provided"""
+
+        for l in self.L:
+            if (not isinstance(l, np.int64)) or l <= 0:
+                raise TypeError("Argument `L` must be a list of positive integers")
+
+        if self.unit_cell.ndim != 2:
+            raise ValueError("'unit_cell' must have ndim==2, as array of primitive vectors.")
+
+        if self.basis.ndim != 2:
+            raise ValueError("'basis' must have ndim==2, as array of initial coordinates.")
+
+        if self.unit_cell.shape[0] != self.unit_cell.shape[1]:
+            raise ValueError("The number of primitive vectors must match their length")
+
+        if not all(isinstance(b, bool) for b in self.boundary_condition):
+            raise ValueError(
+                "Argument 'boundary_condition' must be a bool or a list of bools of same dimensions as the unit_cell"
+            )
+
+        if len(self.boundary_condition) != self.n_dim:
+            raise ValueError(
+                "Argument 'boundary_condition' must be a bool or a list of bools of same dimensions as the unit_cell"
+            )
+
+    def _identify_neighbours(self, cutoff):
+        r"""Identifies the connections between lattice points and returns the unique connections
+        based on the neighbour_order"""
+
+        tree = cKDTree(self.lattice_points)
+        indices = tree.query_ball_tree(tree, cutoff)
+        unique_pairs = set()
+        edges = {}
+        for i, neighbours in enumerate(indices):
+            for neighbour in neighbours:
+                if neighbour != i:
+                    pair = (min(i, neighbour), max(i, neighbour))
+                    if pair not in unique_pairs:
+                        unique_pairs.add(pair)
+                        dist = np.linalg.norm(
+                            self.lattice_points[i] - self.lattice_points[neighbour]
+                        )
+                        # Scale the distance
+                        bin_density = 21621600  # multiple of expected denominators
```

Comment:
This is sorting the edges based on distance, so that we can have it with different neighbour order. This number is a multiple of some expected demoninators, 2,3,5,7 etc. so that we can scale it up for an easier comparison.

Reply:
Any edge-cases side-effect for hardcoding this number? Also, why is it defined inside the loop and not outside?

Reply:
Not that I can think of or come across so far, but any suggestions to get rid of it are welcome as well. Moved it outside the loop.